engine/table/src/main/java/io/deephaven/engine/table/impl/select/analyzers/SelectAndViewAnalyzer.java

```diff
@@ -343,110 +326,449 @@ private static boolean hasConstantValue(final SelectColumn sc) {
         return false;
     }
 
-    private static boolean shouldPreserve(final SelectColumn sc) {
-        // we already know sc is a SourceColumn or switches to a SourceColumn
-        final ColumnSource<?> sccs = sc.getDataView();
-        return sccs instanceof InMemoryColumnSource && ((InMemoryColumnSource) sccs).isInMemory()
+    private static boolean shouldPreserve(
+            final SelectColumn sc,
+            final ColumnSource<?> columnSource) {
+        return columnSource instanceof InMemoryColumnSource && ((InMemoryColumnSource) columnSource).isInMemory()
                 && !Vector.class.isAssignableFrom(sc.getReturnedType());
     }
 
-    static final int BASE_LAYER_INDEX = 0;
-    static final int REDIRECTION_LAYER_INDEX = 1;
+    /** The layers that make up this analyzer. */
+    private final Layer[] layers;
 
-    /**
-     * The layerIndex is used to identify each layer uniquely within the bitsets for completion.
-     */
-    private final int layerIndex;
+    /** Whether the result should be flat. */
+    private final boolean flatResult;
 
-    public SelectAndViewAnalyzer(int layerIndex) {
-        this.layerIndex = layerIndex;
-    }
+    private final BitSet requiredLayers = new BitSet();
+    private final BitSet remainingLayers = new BitSet();
 
-    int getLayerIndex() {
-        return layerIndex;
+    private SelectAndViewAnalyzer(
+            final Layer[] layers,
+            final boolean flatResult) {
+        this.layers = layers;
+        this.flatResult = flatResult;
+        for (final Layer layer : layers) {
+            if (layer.hasRefreshingLogic()) {
+                requiredLayers.set(layer.getLayerIndex());
+            } else {
+                this.layers[layer.getLayerIndex()] = null;
+            }
+        }
     }
 
-    /**
-     * Set the bits in bitset that represent the base layer and optional redirection layer. No other jobs can be
-     * executed until all of these bits are set.
-     *
-     * @param bitset the bitset to manipulate.
-     */
-    abstract void setBaseBits(BitSet bitset);
+    public final static class AnalyzerContext {
+
+        /** The analyzer that we are building. */
+        private final List<Layer> layers = new ArrayList<>();
+        /** The sources that are available to the analyzer, including parent columns. */
+        private final Map<String, ColumnSource<?>> allSources = new LinkedHashMap<>();
+        /** The sources that are available to the analyzer, including parent columns, in result key space. */
+        private final Map<String, ColumnSource<?>> allSourcesInResultKeySpace;
+        /** The sources that are explicitly defined to the analyzer, including preserved parent columns. */
+        private final Map<String, ColumnSource<?>> selectedSources = new HashMap<>();
```

Comment:
We were discussing if the new->selected transition is good.
Also, if published->preserved or something makes more sense.